# Task 2.1 — One Knob at a Time (15 min)
Pick **ONE**: `max_depth ∈ {2, 4, 8, 12}` **or** `learning_rate ∈ {0.01, 0.1, 0.3}`.
Retrain per value, fill the table, then **one sentence**: what changed + your hypothesis why.

In [2]:
import sys, pathlib
p = pathlib.Path.cwd()
while not (p / 'src').exists():
    p = p.parent
sys.path.insert(0, str(p / 'src')); REPO = p
import warnings; warnings.filterwarnings('ignore')

In [4]:
import sys, pathlib
REPO = pathlib.Path(r"e:\AUBG\5th semester\ML\ml-project")
sys.path.insert(0, str(REPO / 'src'))
import warnings; warnings.filterwarnings('ignore')

from course_utils import data
from xgboost import XGBClassifier
from sklearn.metrics import roc_auc_score

df = data.load_processed('telco_clean')
X, y = data.prep_features(df)
S = data.make_splits(X, y)

def run(**params):
    base = dict(n_estimators=200, max_depth=4, learning_rate=0.1,
                eval_metric='auc', random_state=42)
    base.update(params)
    m = XGBClassifier(**base).fit(S['X_train'], S['y_train'])
    return (roc_auc_score(S['y_train'], m.predict_proba(S['X_train'])[:, 1]),
            roc_auc_score(S['y_val'], m.predict_proba(S['X_val'])[:, 1]))

# example: run(max_depth=2)

In [5]:
# your experiment here — loop over YOUR chosen knob and print train/val AUC per value
for d in [2, 4, 8, 12]:
    tr, va = run(max_depth=d)
    print(f"max_depth={d}: train AUC={tr:.3f}, val AUC={va:.3f}, gap={tr-va:.3f}")

max_depth=2: train AUC=0.871, val AUC=0.850, gap=0.022
max_depth=4: train AUC=0.921, val AUC=0.842, gap=0.079
max_depth=8: train AUC=0.993, val AUC=0.833, gap=0.160
max_depth=12: train AUC=0.999, val AUC=0.823, gap=0.176


## Results
| value | train AUC | val AUC | gap
|---|---|---|---|
|  2| 0.871 | 0.850 |0.022
|  4| 0.921 | 0.842 |0.079
|  8| 0.993 | 0.833 |0.160
| 12| 0.999 | 0.823 |0.176

**What changed & why (1 sentence):**
As max_depth increases from 2 to 12 , train AUC rises from 0.871 to 0.999.
However, the val AUC falls from 0.850 to 0.823 with gap growing from 0.02 to 0.18.
This is due to the fact that deeper trees split the data into more specific groups while training which could
include noise. Instead of generalizing patterns that will also work for new customers, it overfits. 